### ISSUE TITLE
Clarification on how a custom adapter sends a TurnContext to the botkit pipeline

### ISSUE THREAD

**Issue by frank-fajardo-a142179** [MicrosoftDocs/bot-docs#2065]
This is really a question.

The documentation [The bot middleware pipeline](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-concept-middleware?view=azure-bot-service-4.0#the-bot-middleware-pipeline) says:

> For each activity, the adapter calls middleware in the order in which you added it. The adapter passes in the context object for the turn and a next delegate, and the middleware calls the delegate to pass control to the next middleware in the pipeline.

However, I'm unable to grasp how a custom adapter can call the first middleware in the bot pipeline. Do you have examples? 
My custom adapter gets notification via web sockets from the messaging channel/platform when a new messages is sent. So my adapter does not have a `processActivity()` similar to a webhook adapter. On the other hand, the web socket connection is managed by the messaging platform's SDK, so my custom adapter does not have a `createSocketServer()`. Both `processActivity()` and `createSocketServer()` methods in the examples I've seen have this code, where `logic` which is received as input, presumably from the botkit:
```
this.runMiddleware(context, logic)
```

So I'm unsure how to do it in my case.

Am I supposed to add `processActivity(context, logic)` on my custom adapter which does the above code (e.g. `this.runMiddleware(context, logic)`), and then do this when the controller is ready...?
```
controller.ready(() => {
  ...
  myCustomBotAdapter.processActivity(controller.handleTurn.bind(controller));
});
```

Please help. Thanks

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 5c8000f6-59aa-472e-d99b-96356084ff4e
* Version Independent ID: 7219b726-fce3-958f-5084-fcefeef659d5
* Content: [Middleware - Bot Service](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-concept-middleware?view=azure-bot-service-4.0)
* Content Source: [articles/v4sdk/bot-builder-concept-middleware.md](https://github.com/MicrosoftDocs/bot-docs/blob/live/articles/v4sdk/bot-builder-concept-middleware.md)
* Service: **bot-service**
* GitHub Login: @ivorb
* Microsoft Alias: **kamrani**

**Comment by JonathanFingold** [comment]
@frank-fajardo-a142179 - asking a dev for information.

**Comment by frank-fajardo-a142179** [comment]
I'm answering my own question. 

What I said above is the correct way.


My adapter has a `listenToEvent()` method which binds handlers to messaging events from the messaging platform. I modified this method to accept an argument `logic` of type `(context: TurnContext) => Promise<void>`. I stored that logic in the adapter, like this:
```lang-ts
this.botkitTurnHandler = logic;
```
Then when my adapter gets messaging events from the messaging platform, it creates a `TurnContext` object for that event, and passes that context to the Botkit pipeline like this:
```lang-ts
this.runMiddleware(context, this.botkitTurnHandler);
```

Then to make sure my adapter's `listenToEvents()` get the Botkit `handlerTurn()` function, I do this:
```lang-ts
const controller: Botkit = new Botkit({...});
controller.ready(() => {
  ...
  const turnHandler = controller.handleTurn.bind(controller);
  myAdapter.listenToEvents(turnHandler );
});
```
This is based on documents/comments made [here](https://github.com/howdyai/botkit/blob/93650237bd19a8cc8aac379436c30f8bb0116c14/packages/botkit/src/core.ts#L705) and [here](https://github.com/howdyai/botkit/blob/93650237bd19a8cc8aac379436c30f8bb0116c14/packages/botbuilder-adapter-web/src/web_adapter.ts#L55)
